src/NUnitFramework/framework/Assert.Exceptions.Async.cs

```
@@ -0,0 +1,281 @@
+        /// <param name="message">The message that will be displayed on failure</param>
+        /// <param name="args">Arguments to be used in formatting the message</param>
+        public static Task<TActual> CatchAsync<TActual>(AsyncTestDelegate code, string message, params object[] args) where TActual : System.Exception
+        {
+            return Cast<TActual>(ThrowsAsync(new InstanceOfTypeConstraint(typeof (TActual)), code, message, args));
+        }
+
+        /// <summary>
+        /// Verifies that an async delegate throws an exception of a certain Type
+        /// or one derived from it when called and returns it.
+        /// </summary>
+        /// <param name="code">A TestDelegate</param>
+        public static Task<TActual> CatchAsync<TActual>(AsyncTestDelegate code) where TActual : System.Exception
+        {
+            return Cast<TActual>(ThrowsAsync(new InstanceOfTypeConstraint(typeof (TActual)), code));
+        }
+
+        #endregion
+
+        #region DoesNotThrowAsync
+
+        /// <summary>
+        /// Verifies that an async delegate does not throw an exception
+        /// </summary>
+        /// <param name="code">A TestDelegate</param>
+        /// <param name="message">The message that will be displayed on failure</param>
+        /// <param name="args">Arguments to be used in formatting the message</param>
+        public static Task DoesNotThrowAsync(AsyncTestDelegate code, string message, params object[] args)
+        {
+            Exception caughtException;
+            try
+            {
+                var task = code();
+                return task.ContinueWith(t =>
+                {
+                    // Case where code() either
+                    // - succeeded (synchronously or asynchronously)
+                    // - failed asynchronously
+                    var ex = t.IsFaulted ? t.Exception.InnerException : null;
+                    Assert.That(ex, new ThrowsNothingConstraint(), message, args);
```

@CharliePoole, you're right, I don't need to return the exception, so I should drop the `return ex` part. But I still need an assertion that will produce the right message (i.e. "Expected: No Exception to be thrown/But was: ..."). That's why I used `ThrowsNothingConstraint`. Should I create a specific `Constraint`?